docs/RFCS/20210722_on_update_and_row_rehoming.md

```diff
@@ -0,0 +1,250 @@
+SELECT quantity_on_hand FROM inventories;
+> 1
+UPDATE inventories SET product_id = 2 WHERE warehouse_id = 1;
+SELECT quantity_on_hand FROM inventories;
+> 1
+ALTER TABLE inventories ALTER COLUMN quantity_on_hand SET ON UPDATE 50;
+UPDATE inventories SET product_id = 3 WHERE warehouse_id = 1;
+SELECT quantity_on_hand FROM inventories;
+> 50
+UPDATE inventories SET quantity_on_hand = 100;
+-- how to UPDATE keeping current value for quantity
+UPDATE inventories i SET (product_id, quantity_on_hand) = (4, i.quantity_on_hand);
+SELECT quantity_on_hand FROM inventories;
+> 100
+```
+
+## Row Rehoming
+
+To implement row rehoming, an ON UPDATE constraint will be added to the
+crdb_region column in REGIONAL BY ROW tables. This will make it so that when an
+UPDATE occurs that does not touch `crdb_region`, the `gateway_region()` default
+clause will be evaluated and applied.
+
+This has the potential to cause thrashing if UPDATEs happen from many different
+regions, so this behavior will be gated behind a cluster setting,
+`sql.defaults.auto_rehome.enabled`.
+
+If the user wishes to apply this behavior to just one of their tables as opposed
+to applying a cluster-wide setting, they can create the REGIONAL BY ROW table
+and alter its `crdb_region` column with an ON UPDATE expression , or they can specify
+ON UPDATE on their `crdb_internal_region` column and specify the column with
+REGIONAL BY ROW AS.
+
+Example of creating an auto-rehoming table:
+
+```sql
+CREATE TABLE test (
+  p int,
+  region crdb_internal_region
+    DEFAULT default_to_database_primary_region(gateway_region())
+    ON UPDATE default_to_database_primary_region(gateway_region())
+)
+  LOCALITY REGIONAL BY ROW AS region;
+```
+
+## Drawbacks
+
+The primary drawback of this option for rehoming is that ON UPDATE expressions
+cannot reference any other columns, so there is no way to implement any
+heuristic for choosing when to rehome a row. For example, a heuristic like “only
+rehome after 5 sequential UPDATEs” would be impossible to implement. Depending
+on the client workload, a heuristic like this may decrease thrashing,
+particularly if the vast majority of UPDATEs come from a single region but some
+occasionally come from other regions.
+
+## Rationale and Alternatives
+
+### ON UPDATE
+
+It's worth noting that the ON UPDATE syntax is heavily inspired by [MySQL's ON
+UPDATE
+CURRENT_TIMESTAMP](https://dev.mysql.com/doc/refman/8.0/en/timestamp-initialization.html)
+syntax. In terms of functionality, however, this proposal is distinct in that it
+accepts arbitrary expressions while the only MySQL usage of `ON UPDATE` is `ON
+UPDATE CURRENT_TIMESTAMP`. This proposal proposes a more general feature so that
+more workloads can be enabled. Because of the [special syntax
+form](https://dev.mysql.com/doc/refman/8.0/en/timestamp-initialization.html) for
+`current_timestamp()`, `ON UPDATE CURRENT_TIMESTAMP` is valid usage under this
```

Comment:
how does this compare to using a computed column?

we don't currently allow non-deterministic operators in a computed column definition. e.g.
```
root@:26257/defaultdb> create table t2 (a int primary key, b timestamptz as (current_timestamp()) virtual);
ERROR: current_timestamp(): context-dependent operators are not allowed in computed column
```

so what's different about ON UPDATE that lets us use `current_timestamp()` here?

Reply:
I would think about ON UPDATE as closer to a DEFAULT statement than a computed column; context-dependent operations are allowed, and they will be evaluated using the context present at UPDATE time just like DEFAULTs.

The reason for this difference is that computed columns are meant to derive their value solely from the state of the other columns in the row whereas ON UPDATE is _not_ meant to use the state of the other rows. It's only supposed to use the context.

Reply:
Yep; Just added.